File: leetcode/medium/150_reverse_polish_notation/reverse_polish_notation.py
from typing import List


class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        for token in tokens:
            if token == "+":
                a = stack.pop()
                b = stack.pop()
                stack.append(b + a)
            elif token == "-":
                a = stack.pop()
                b = stack.pop()
                stack.append(b - a)
            elif token == "*":
                a = stack.pop()
                b = stack.pop()
                stack.append(b * a)
            elif token == "/":
                a = stack.pop()
                b = stack.pop()
                stack.append(int(b / a))
            else:
                stack.append(int(token))
        return stack[0]


print(Solution().evalRPN(["10", "6", "9", "3", "+",
      "-11", "*", "/", "*", "17", "+", "5", "+"]))

File: leetcode/medium/150_reverse_polish_notation/reverse_polish_notation.cpp
#include <string>
#include <vector>

class Solution {
public:
  int evalRPN(std::vector<std::string> &tokens) {
    std::vector<int> stack;
    for (auto token : tokens) {
      if (token == "+") {
        int val1 = stack.back();
        stack.pop_back();
        int val2 = stack.back();
        stack.pop_back();
        stack.push_back(val2 + val1);
      } else if (token == "-") {
        int val1 = stack.back();
        stack.pop_back();
        int val2 = stack.back();
        stack.pop_back();
        stack.push_back(val2 - val1);
      } else if (token == "*") {
        int val1 = stack.back();
        stack.pop_back();
        int val2 = stack.back();
        stack.pop_back();
        stack.push_back(val2 * val1);
      } else if (token == "/") {
        int val1 = stack.back();
        stack.pop_back();
        int val2 = stack.back();
        stack.pop_back();
        stack.push_back(val2 / val1);
      } else {
        stack.push_back(stoi(token));
      }
    }
    return stack[0];
  }
};
